### PR TITLE
SIDM-3089 Fix first time execution errors: 'ResourceGroupNotFound' & ''Cannot find ServerFarm'

### PR DESCRIPTION
**Pull request by henrydobson**
### Change description ###

Pipelines using the cnp-module-webapps must run 3 times to pass.

During Terraform execution, two failures will always occur for first time deployments before a successful execution.

1. `Resource group <name> could not be found`
2. `Cannot find ServerFarm with name <name>`

This is due to incorrect dependencies. By using the resource reference instead of the local within the Terraform resources, this enforces implicit dependencies for the RG resolving `ResourceGroupNotFound` errors.

By adding the `dependsOn` statement for the serverfarm deployment in the ARM template, this removes the 'Cannot find ServerFarm with name <name>' error.

```bash
# Execution 1
Error: Error applying plan:
2 error(s) occurred:
* module.idam-web-admin.azurerm_template_deployment.tmprofile: 1 error(s) occurred:
* azurerm_template_deployment.tmprofile: Error creating deployment: resources.DeploymentsClient#CreateOrUpdate: Failure sending request: StatusCode=404 -- Original Error: Code="ResourceGroupNotFound" Message="Resource group '<name>' could not be found."

# Execution 2
Error: Error applying plan:
1 error(s) occurred:
* module.idam-web-admin.azurerm_template_deployment.app_service_site: 1 error(s) occurred:
* azurerm_template_deployment.app_service_site: Error waiting for deployment: Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." Details=[{"code":"NotFound","message":"{\r\n  \"Code\": \"NotFound\",\r\n  \"Message\": \"Cannot find ServerFarm with name <name>.\",\r\n  \"Target\": null,\r\n  \"Details\": [\r\n    {\r\n      \"Message\": \"Cannot find ServerFarm with name <name>.\"\r\n    },\r\n    {\r\n      \"Code\": \"NotFound\"\r\n    },\r\n    {\r\n      \"ErrorEntity\": {\r\n        \"ExtendedCode\": \"51004\",\r\n        \"MessageTemplate\": \"Cannot find {0} with name {1}.\",\r\n        \"Parameters\": [\r\n          \"ServerFarm\",\r\n          \"<name>\"\r\n        ],\r\n        \"Code\": \"NotFound\",\r\n        \"Message\": \"Cannot find ServerFarm with name <name>\"\r\n      }\r\n    }\r\n  ],\r\n  \"Innererror\": null\r\n}"}]
```

Resource groups should be created outside of the module declaration and passed into the module as a param but a conditional resource has been retained for backwards compatibility.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by henrydobson** [comment]
I have discovered an issue with execution when overriding the asp_name. I have marked the PR as WIP until resolved.

**Comment by henrydobson** [comment]
WIP now complete for cases when overriding the asp_name. Ready for review and merge.

**Comment by naikajah** [comment]
@henrydobson please do not merge this PR. Currently work is being done to update this module in such a way that it facilitates the clean up of ASE's when things will get migrated to AKS.

File: examples/main.tf [hunk 1]
```diff
@@ -0,0 +1,74 @@
+provider "azurerm" {
+  version = "1.22.1"
```
Comment (timja) [inline examples/main.tf]: no example should be using a provider anywhere near this old

File: examples/main.tf [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    "Team Name" = "IDAM"
+  }
+}
+
+# Both default_asp_name_and_rg and params_asp_name_and_rg must be included to 
+# ensure that the dependencies and references work correctly in both examples.
+# A common error is to include a new dependsOn resourceId without the RG
+# parameter, which will work for default_asp_name_and_rg but fail on
+# params_asp_name_and_rg.
+module "default_asp_name_and_rg" {
```
Comment (timja) [inline examples/main.tf]: not sure what you're trying to achieve here, the second example is what people should be doing by default

File: main.tf [hunk 3]
```diff
@@ -24,6 +24,12 @@ resource "azurerm_resource_group" "rg" {
   tags = "${merge(var.common_tags,
     map("lastUpdated", "${timestamp()}")
     )}"
+  
+  # On creation, resource group is not ready without delay.
```
Comment (timja) [inline main.tf]: I've never seen this before =/, is there an upstream bug with terraform?
Given we're going to be removing this resource I would leave this change out

File: main.tf [hunk 4]
```diff
@@ -34,6 +40,12 @@ resource "azurerm_resource_group" "rg2" {
   tags = "${merge(var.common_tags,
     map("lastUpdated", "${timestamp()}")
     )}"
+    
+  # On creation, resource group is not ready without delay.
+  provisioner "local-exec" {
```
Comment (timja) [inline main.tf]: Pretty sure this resource group isn't even used? so I would remove this change

File: main.tf [hunk 5]
```diff
@@ -47,7 +59,7 @@ resource "azurerm_application_insights" "appinsights" {
 
   name                = "${var.product}-appinsights-${var.env}${var.deployment_target}"
   location            = "${var.appinsights_location}"
-  resource_group_name = "${local.resource_group_name}"
+  resource_group_name = "${azurerm_resource_group.rg.name}"
```
Comment (timja) [inline main.tf]: please revert this change, we're in the progress of migrating applications to always passing in a resource group to the module, as we're decommissioning this module as part of moving to Kubernetes

REQ

File: main.tf [hunk 6]
```diff
@@ -76,7 +88,7 @@ resource "azurerm_template_deployment" "app_service_site" {
   count               = "${var.enable_ase}"
   template_body       = "${data.template_file.sitetemplate.rendered}"
   name                = "${var.product}-${var.env}${var.deployment_target}-webapp"
-  resource_group_name = "${local.resource_group_name}"
+  resource_group_name = "${azurerm_resource_group.rg.name}"
```
Comment (timja) [inline main.tf]: please revert this change, we're in the progress of migrating applications to always passing in a resource group to the module, as we're decommissioning this module as part of moving to Kubernetes

REQ

File: main.tf [hunk 7]
```diff
@@ -110,7 +122,7 @@ resource "azurerm_template_deployment" "app_service_ssl" {
 
   template_body       = "${data.template_file.ssltemplate.rendered}"
   name                = "${var.product}-${var.env}${var.deployment_target}-cert"
-  resource_group_name = "${local.resource_group_name}"
+  resource_group_name = "${azurerm_resource_group.rg.name}"
```
Comment (timja) [inline main.tf]: please revert this change, we're in the progress of migrating applications to always passing in a resource group to the module, as we're decommissioning this module as part of moving to Kubernetes

REQ

File: trafficman.tf [hunk 8]
```diff
@@ -5,7 +5,7 @@ data "template_file" "tmtemplate" {
 resource "azurerm_template_deployment" "tmprofile" {
   template_body       = "${data.template_file.tmtemplate.rendered}"
   name                = "${var.product}-${var.env}-tm"
-  resource_group_name = "${local.resource_group_name}"
+  resource_group_name = "${azurerm_resource_group.rg.name}"
```
Comment (timja) [inline trafficman.tf]: please revert this change, we're in the progress of migrating applications to always passing in a resource group to the module, as we're decommissioning this module as part of moving to Kubernetes

REQ

File: examples/main.tf (hunk 1 above)
Comment (henrydobson) [reply examples/main.tf]: This was used to remain inline with the version used for IDAM web apps. Assumed this was a version constraint. 

Update

File: examples/main.tf (hunk 2 above)
Comment (henrydobson) [reply examples/main.tf]: The second example `params_asp_name_and_rg` is how IDAM is implemented however the module supports no ASP and RG name parameters, which affects the successful execution of the ARM template if the `dependsOn` resource is not referenced using the `resourceId` parameter. 

Example: https://github.com/hmcts/cnp-module-webapp/blob/5d7fe9a1ad743efb4581bc50e99661041576b2ec/templates/asp-app.json#L157

This examples directory simply tests both module configurations for a more complete test.

File: main.tf (hunk 4 above)
Comment (henrydobson) [reply main.tf]: I had this thought also. 

Remove

File: main.tf (hunk 5 above)
Comment (henrydobson) [reply main.tf]: https://github.com/hmcts/cnp-module-webapp/pull/156#discussion_r322741012

File: main.tf (hunk 6 above)
Comment (henrydobson) [reply main.tf]: https://github.com/hmcts/cnp-module-webapp/pull/156#discussion_r322741012

File: trafficman.tf (hunk 8 above)
Comment (henrydobson) [reply trafficman.tf]: https://github.com/hmcts/cnp-module-webapp/pull/156#discussion_r322741012

File: main.tf (hunk 3 above)
Comment (timja) [reply main.tf]: if you pass a resource group in from outside the module it will have the same affect

**Comment by timja** [comment]
> These changes were made to remove the deployment issues in PR pipelines for IDAM web applications. These issues were disruptive and whilst IDAM is working towards AKS, it's a reality that ASE environments will be used for a long enough time for the behaviour of this module to cause a fair amount of wasted time troubleshooting and re-running pipelines.
> 
> If this module is being deprecated, further review is not required as we are only going to be using this until we are ready for AKS.

it's still a production module so if you take a look at my comments we can get this merged you won't be able to use this branch after ~2 weeks when non master branches are getting blocked

File: templates/asp-app.json [hunk 9]
```diff
@@ -153,6 +153,9 @@
         "environment": "[parameters('env')]",
         "Team Name": "[parameters('teamName')]"
       },
+      "dependsOn": [
```
Comment (timja) [inline templates/asp-app.json]: this change should be fine to be mainlined if you separate it out to a different PR

File: main.tf [hunk 10]
```diff
@@ -1,6 +1,6 @@
 locals {
   default_resource_group_name = "${var.product}-${var.env}${var.deployment_target}"
-  resource_group_name         = "${var.resource_group_name != "" ? var.resource_group_name : local.default_resource_group_name}"
+  resource_group_name         = "${coalesce(var.resource_group_name, local.default_resource_group_name)}"
```
Comment (timja) [inline main.tf]: this change should be fine to be mainlined if you separate it out to a different PR

File: main.tf [hunk 11]
```diff
@@ -9,7 +9,7 @@ locals {
   }
 
   asp_name = "${var.asp_name != "null" ? var.asp_name : local.default_resource_group_name}"
-  asp_rg   = "${var.asp_rg != "null" ? var.asp_rg : local.default_resource_group_name}"
+  asp_rg   = "${var.asp_rg != "null" ? var.asp_rg : local.resource_group_name}"
```
Comment (timja) [inline main.tf]: this change should be fine to be mainlined if you separate it out to a different PR

File: main.tf (hunk 10 above)
Comment (henrydobson) [reply main.tf]: If the intent is to declare a Resource Group externally to the module and pass it in as a module parameter then this change is can be removed in favour of:
```suggestion
```

File: main.tf (hunk 11 above)
Comment (henrydobson) [reply main.tf]: If the intent is to declare a Resource Group externally to the module and pass it in as a module parameter then this change is can be removed in favour of:

```suggestion
  asp_rg   = "${var.asp_rg}"
```

File: main.tf (hunk 10 above)
Comment (timja) [reply main.tf]: unfortunately there's backwards compatibility to keep and we can't really make that change as it would be breaking

File: main.tf (hunk 11 above)
Comment (timja) [reply main.tf]: unfortunately there's backwards compatibility to keep and we can't really make that change as it would be breaking

File: README.md [hunk 12]
```diff
@@ -45,16 +56,14 @@ module "frontend" {
 	capacity             = "${var.capacity}"
 	is_frontend          = true
 	asp_name             = "${var.product}-${var.env}"
-	asp_rg               = "${var.product}-shared-infrastructure-${var.env}"
+  asp_rg 							 = "${azurerm_resource_group.default.name}" # RG param
```
Comment (timja) [inline README.md]: looks like tabs and spaces have been mixed here, formatting looks weird

**Comment by henrydobson** [comment]
Just reverting some changes to retain the backwards compatibility

File: main.tf [hunk 13]
```diff
@@ -1,41 +1,19 @@
...
     WEBSITE_LOCAL_CACHE_OPTION   = "${var.website_local_cache_sizeinmb == "0" ? "Never" : "Always"}"
     WEBSITE_LOCAL_CACHE_SIZEINMB = "${var.website_local_cache_sizeinmb}"
   }
 
-  asp_name = "${var.asp_name != "null" ? var.asp_name : local.default_resource_group_name}"
-  asp_rg   = "${var.asp_rg != "null" ? var.asp_rg : local.default_resource_group_name}"
-  sp_name  = "${var.env != "preview" ? local.asp_name : local.default_resource_group_name}"
-  sp_rg    = "${var.env != "preview" ? local.asp_rg : local.default_resource_group_name}"
+  asp_name = "${var.asp_name != "null" ? var.asp_name : local.asp_rg}"
+  asp_rg   = "${var.asp_rg}"
```
Comment (timja) [inline main.tf]: doesn't this need to fall back to the default_resource_group_name?

File: variables.tf [hunk 14]
```diff
@@ -121,8 +121,7 @@ variable "common_tags" {
 }
 
 variable "asp_rg" {
-  description = "Name of the resource group where the asp specified in asp_name resides"
-  default     = "null"
```
Comment (timja) [inline variables.tf]: there's specific handling in the arm template for this value, I would leave it in as is...

File: main.tf (hunk 10 above)
Comment (henrydobson) [reply main.tf]: @timja I think it's not really achievable to resolve these execution issues, support an external resource group declaration and retain backward compatibility.

Terraform 0.11.x doesn't really have sufficient expressions and logic to do this and obviously 0.12.x is a breaking change in itself.

I have tried a few theories and below are my findings.

Use cases:
- ⚠ cnp-module-webapps defined without `asp_rg` (No external RG)
```
A conditional resource in the module will create a resource group with default name.
NOTE: This does NOT run first time. Due to limitations in Terraform 0.11.x this implementation will work but will first fail with a 404 ResourceGroupNotFound error.
```
- ❗️cnp-module-webapps defined with `asp_rg` (No external RG)
```
The module will assume that the resource group exists and will use `local.asp_rg` for all resource group references. 
NOTE: This will fail if the resource group does not exist!
```
- ❗️cnp-module-webapps defined with `asp_rg` and resource group exists already due to a previous execution (No external RG)
```
The module will plan destruction of the resource group and cause the module to fail.
```
- ✅ cnp-module-webapps defined with `asp_rg = ${azurerm_resource_group.default.name}` (External RG)
```
The module will plan destruction of the resource group and cause the module to fail.
```

Due to this and the planned deprecation of this module, I'm going to have to suggest that we close this PR for good and submit a PR for allowing the use of this branch in the Terraform whitelist.  

Do you agree?

Comment (timja) [reply main.tf]: Somewhat, but when we move to an external resource group we do a state migration (https://github.com/hmcts/terraform-state-move)

**Comment by henrydobson** [comment]
Closing see https://github.com/hmcts/cnp-jenkins-config/pull/320
